contracts/OtokenFactory.sol

```diff
@@ -0,0 +1,185 @@
+pragma solidity =0.6.10;
+
+import {OtokenSpawner} from "./OtokenSpawner.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
+import {OtokenInterface} from "./interfaces/OtokenInterface.sol";
+import {WhitelistInterface} from "./interfaces/WhitelistInterface.sol";
+
+/**
+ * SPDX-License-Identifier: UNLICENSED
+ * @title A factory for opyn tokens
+ * @author Opyn
+ * @notice Create new otokens and keep track of all created tokens.
+ * @dev Calculate contract address before each creation with CREATE2
+ * and deploy eip-1167 minimal proxies for otoken logic contract.
+ */
+contract OtokenFactory is OtokenSpawner {
+    /// @notice The Opyn AddressBook contract that records addresses of whitelist module and otoken impl address. */
+    address public addressBook;
+
+    /// @dev An array of all created otokens */
+    address[] private _otokens;
+
+    /// @dev A mapping from parameters hash to its deployed address
+    mapping(bytes32 => address) private _idToAddress;
+
+    constructor(address _addressBook) public {
+        addressBook = _addressBook;
+    }
+
+    /* Event: OtokenCreated - Emitted when factory create a new Option. */
+    event OtokenCreated(
+        address indexed tokenAddress,
+        address indexed creator,
+        address underlying,
+        address strike,
+        address collateral,
+        uint256 strikePrice,
+        uint256 expiry,
+        bool isPut
+    );
+
+    /**
+     * @notice create new otokens
+     * @dev deploy an eip-1167 minimal proxy with CREATE2 and register it to the whitelist module.
+     * @param _underlyingAsset asset that the option references
+     * @param _strikeAsset asset that the strike price is denominated in
+     * @param _collateralAsset asset that is held as collateral against short/written options
+     * @param _strikePrice strike price with decimals = 18
+     * @param _expiry expiration timestamp in second
+     * @param _isPut is this a put option, if not it is a call
+     * @return newOtoken address of the newly created option
+     */
+    function createOtoken(
+        address _underlyingAsset,
+        address _strikeAsset,
+        address _collateralAsset,
+        uint256 _strikePrice,
+        uint256 _expiry,
+        bool _isPut
+    ) external returns (address newOtoken) {
+        require(_expiry > now, "OtokenFactory: Can't create expired option.");
```

Comment:
might also want to check expiry is a valid unix timestamp

Reply:
@aleone and I were chatting about potentially add more restriction on expiry, but not sure should we have them here or in the Otoken contract
I currently plan to have the max expiry check in the `oToken` init function (the expiry can't be greator than the max uint bokkypoobah tested ), because Otoken contract use the bokkypoobah library.  Should we move that check here as well so we can be more consistent?

Another check to include: 
* (expiry - 28800) mod 86400 == 0, to ensure all the option expires on 08:00 UTC.

@aparnakr  @haythem96

Reply:
I included following 2 checks in 804b7ea
*  all expiry has to be 08:00 UTC
* expiry < 2345/12/31, which is the max DatetimeLibrary supports

Reply:
hmm does it need to be 08:00 UTC? Will we have expiries that are different @aleone?